File: python/1254. Number of Closed Islands.py
class Solution(object):
    def closedIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m, n = len(grid), len(grid[0])
        vis = [[0]*n for _ in range(m)]

        def dfs(x, y):
            vis[x][y] = 1

            ans = True
            for dx, dy in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                nx = x + dx
                ny = y + dy
                if nx<0 or nx>=m or ny<0 or ny>=n:
                    ans = False
                    continue
                if vis[nx][ny]==0 and grid[nx][ny]==0 and dfs(nx, ny) is False:
                    ans = False
            return ans
        
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j]==0 and vis[i][j]==0 and dfs(i, j):
                    res += 1
        return res

File: go/src/leetcode/1254/1254. Number of Closed Islands.go
package main

/*
*
https://leetcode.cn/problems/number-of-closed-islands/description/
*/

func closedIsland(grid [][]int) (tot int) {
	m, n := len(grid), len(grid[0])
	var dfs func(x, y int)
	dfs = func(x, y int) {
		if x < 0 || x >= m || y < 0 || y >= n || grid[x][y] != 0 {
			return
		}
		grid[x][y] = 1
		dfs(x+1, y)
		dfs(x-1, y)
		dfs(x, y+1)
		dfs(x, y-1)
	}
	for i := 0; i < m; i++ {
		dfs(i, 0)
		dfs(i, n-1)
	}
	for j := 0; j < n; j++ {
		dfs(0, j)
		dfs(m-1, j)
	}

	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == 0 {
				tot++
				dfs(i, j)
			}
		}
	}
	return tot
}

func main() {
	closedIsland([][]int{{1, 1, 1, 1, 1, 1, 1, 0}, {1, 0, 0, 0, 0, 1, 1, 0}, {1, 0, 1, 0, 1, 1, 1, 0},
		{1, 0, 0, 0, 0, 1, 0, 1}, {1, 1, 1, 1, 1, 1, 1, 0}})
}
